src/messaging/CredentialApiMessageTypes.ts

```diff
@@ -0,0 +1,133 @@
+    }
+    break
+  }
+  case 'request-attestation': {
+    Credential.verifyDataStructure(body.content.credential)
+    if (body.content.quote) {
+      Quote.validateQuoteSchema(Quote.QuoteSchema, body.content.quote)
+    }
+    break
+  }
+  case 'submit-attestation': {
+    Attestation.verifyDataStructure(body.content.attestation)
+    break
+  }
+  case 'reject-attestation': {
+    if (!isHex(body.content)) {
+      throw new SDKErrors.HashMalformedError()
+    }
+    break
+  }
+  case 'request-credential': {
+    body.content.cTypes.forEach(({ cTypeHash, trustedAttesters, requiredProperties }): void => {
+      DataUtils.verifyIsHex(cTypeHash)
+      trustedAttesters?.forEach((did) => Did.validateUri(did, 'Did'))
+      requiredProperties?.forEach((requiredProps) => {
+        if (typeof requiredProps !== 'string') throw new TypeError('Required properties is expected to be a string')
+      })
+    })
+    break
+  }
+  case 'submit-credential': {
+    body.content.forEach((presentation) => {
+      Credential.verifyDataStructure(presentation)
+      if (!Did.isDidSignature(presentation.claimerSignature)) {
+        throw new SDKErrors.SignatureMalformedError()
+      }
+    })
+    break
+  }
+
+  default:
+    throw new SDKErrors.UnknownMessageBodyTypeError()
+  }
+}
+
+/**
+     * Verifies that the sender of a [[Message]] is also the owner of it, e.g the owner's and sender's DIDs refer to the same subject.
+     *
+     * @param message The [[Message]] object which needs to be decrypted.
+     * @param message.body The body of the [[Message]] which depends on the [[BodyType]].
+     * @param message.sender The sender's DID taken from the [[IMessage]].
+     */
+export function ensureOwnerIsSender({ body, sender }: IMessage): void {
+  switch (body.type) {
+  case 'request-attestation':
+    {
+      const requestAttestation = body
+      if (!Did.isSameSubject(requestAttestation.content.credential.claim.owner, sender)) {
+        throw new SDKErrors.IdentityMismatchError('Claim', 'Sender')
+      }
+    }
+    break
+  case 'submit-attestation':
+    {
+      const submitAttestation = body
+      if (!Did.isSameSubject(submitAttestation.content.attestation.owner, sender)) {
+        throw new SDKErrors.IdentityMismatchError('Attestation', 'Sender')
+      }
+    }
+    break
+  case 'submit-credential':
+    {
+      const submitClaimsForCtype = body
+      submitClaimsForCtype.content.forEach((presentation) => {
+        if (!Did.isSameSubject(presentation.claim.owner, sender)) {
+          throw new SDKErrors.IdentityMismatchError('Claims', 'Sender')
+        }
+      })
+    }
+    break
+  default:
+  }
+}
+
+/**
+     * Checks the message structure and body contents (e.g. Hashes match, ensures the owner is the sender).
+     * Throws, if a check fails.
+     *
+     * @param decryptedMessage The decrypted message to check.
+     */
+export function verify(decryptedMessage: IMessage): void {
```

Comment:
could also be a type guard

Reply:
Same here? Other name or what do you mean?

Reply:
See above. Typeguards narrow a type, the point of this being that if you start with a `IMessage<{type:string, content: unknown}>` you can use the verify function to narrow it to a `IMessage<MessageBody>` type like so:
```ts
const message: IMessage<{type:string, content: unknown}> = decrypt(...)
if(!verify(message)) {
  throw new Error("not a recognised message type")
}
if(message.body.type === 'submit-terms') {
  message.body.content.claim // tsc now knows this must exist despite the initial 'unknown' typing
}
```

I'm realising though that the purpose of making these checks can be better achieved if we'd refactor it to be not a type guard, but to return the message (body) and throw in case of validation errors, so we can do:
```ts
const message: IMessage<{type:string, content: unknown}> = decrypt(...)
const { body } = assertKnownMessage(message)
// body is typed as MessageBody now
if(body.type === 'submit-terms') {
  body.content.claim // this works again
}
```

Reply:
I've seen you added type guards for the individual message types though, which is almost better! There are some validations though that need both the message body and envelope, like the ensureOwnerIsSender. This could be addressed by making those type guards consume the entire message object, something like this:
```ts
// before
export function isSubmitTerms(body: MessageBody): body is ISubmitTerms
// after
export function isSubmitTerms(message: IMessage<{type:string, content: unknown}>): body is IMessage<ISubmitTerms>
```
which let's you do
```ts
const message: IMessage<{type:string, content: unknown}> = decrypt(...)
if(isSubmitTerms(message)) {
 throw new Error('not the message we expected!')
}
message.body.content.claim // this works again
```

The only concern I can see @arty-name raising here is that this pattern swallows info on what is wrong with the message. This is something we could address with `assertIsSubmitTerms(message: IMessage): IMessage<ISubmitTerms>` style message type checkers.